graylog2-server/src/main/java/org/graylog2/rest/resources/streams/StreamResource.java

```diff
@@ -189,6 +209,41 @@ private boolean ensureUserHasPermissionsForStream(User user, String id) throws V
         return permissionsChanged;
     }
 
+    @GET
+    @Timed
+    @Path("/paginated")
+    @ApiOperation(value = "Get a paginated list of streams")
+    @Produces(MediaType.APPLICATION_JSON)
+    public StreamPageListResponse getPage(@ApiParam(name = "page") @QueryParam("page") @DefaultValue("1") int page,
+        @ApiParam(name = "per_page") @QueryParam("per_page") @DefaultValue("50") int perPage,
+        @ApiParam(name = "query") @QueryParam("query") @DefaultValue("") String query,
+        @ApiParam(name = "sort",
+                value = "The field to sort the result on",
+                required = true,
+                allowableValues = "title,description")
+        @DefaultValue(StreamImpl.FIELD_TITLE) @QueryParam("sort") String sort,
+        @ApiParam(name = "order", value = "The sort direction", allowableValues = "asc, desc")
+        @DefaultValue("asc") @QueryParam("order") String order) {
+
+        SearchQuery searchQuery;
+        try {
+            searchQuery = searchQueryParser.parse(query);
+        } catch (IllegalArgumentException e) {
+            throw new BadRequestException("Invalid argument in search query: " + e.getMessage());
+        }
+        final Predicate<StreamDTO> permissionFilter = streamDTO -> isPermitted(RestPermissions.STREAMS_READ, streamDTO.id());
+        final PaginatedList<StreamDTO> result = paginatedStreamService
+                .findPaginated(searchQuery, permissionFilter, page, perPage, sort, order);
+        final List<StreamDTO> streams = result.stream().map(streamDTO -> {
+            List<StreamRule> rules = streamRuleService.loadForStreamId(streamDTO.id());
```

Comment:
Instead of loading rules for each stream separately, we should use `StreamRuleService#loadForStreamIds` so we are generating 2 queries instead of n+1.